File: legacy/graphql-java/src/main/java/com/example/demo/gql/DataLoaders.java
package com.example.demo.gql;

import com.example.demo.gql.types.Author;
import com.example.demo.gql.types.Comment;
import com.example.demo.service.AuthorService;
import com.example.demo.service.PostService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.dataloader.BatchLoader;
import org.dataloader.DataLoader;
import org.dataloader.MappedBatchLoader;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.CompletableFuture;

@Slf4j
@RequiredArgsConstructor
@Component
public class DataLoaders {
    final AuthorService authorService;
    final PostService postService;

    public DataLoader<String, Author> authorsLoader() {
        BatchLoader<String, Author> batchLoader =
                (List<String> keys) -> CompletableFuture.supplyAsync(() -> authorService.getAuthorByIdIn(keys));
        return DataLoader.newDataLoader(batchLoader);
    }

    public DataLoader<String, List<Comment>> commentsLoader() {
        MappedBatchLoader<String, List<Comment>> batchLoader = (Set<String> keys) -> {
            List<Comment> comments = this.postService.getCommentsByPostIdIn(keys);

            Map<String, List<Comment>> mappedComments = new HashMap<>();
            keys.forEach(
                    k -> mappedComments.put(k, comments
                            .stream()
                            .filter(c -> c.getPostId().equals(k)).toList())
            );
            log.info("mapped comments: {}", mappedComments);
            return CompletableFuture.supplyAsync(() -> mappedComments);
        };
        return DataLoader.newMappedDataLoader(batchLoader);
    }
}

File: graphql-kotlin/src/main/kotlin/com/example/demo/gql/datafetchers/DataLoaders.kt
package com.example.demo.gql.datafetchers

import com.example.demo.service.AuthorService
import com.example.demo.service.PostService
import com.example.demo.gql.types.Author
import com.example.demo.gql.types.Comment
import com.expediagroup.graphql.dataloader.KotlinDataLoader
import graphql.GraphQLContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.asCoroutineDispatcher
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.future.future
import org.dataloader.DataLoader
import org.dataloader.DataLoaderFactory
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.util.*
import java.util.concurrent.Executors

val loaderScope = CoroutineScope(Executors.newCachedThreadPool().asCoroutineDispatcher())

@Component
class AuthorsDataLoader(val authorService: AuthorService) : KotlinDataLoader<UUID, Author> {
    companion object {
        private const val NAME = "AuthorsDataLoader"
    }

    override val dataLoaderName = NAME
    override fun getDataLoader(graphQLContext: GraphQLContext): DataLoader<UUID, Author> {
        return DataLoaderFactory.newDataLoader { keys, _ ->
            loaderScope.future {
                authorService.getAuthorByIdIn(keys).toList()
            }
        }
    }
}


@Component
class CommentsDataLoader(val postService: PostService) : KotlinDataLoader<UUID, List<Comment>> {
    companion object {
        private val log = LoggerFactory.getLogger(CommentsDataLoader::class.java)
        private const val NAME = "CommentsDataLoader"
    }

    override val dataLoaderName = NAME
    override fun getDataLoader(graphQLContext: GraphQLContext): DataLoader<UUID, List<Comment>> {
        return DataLoaderFactory.newMappedDataLoader { keys, _ ->
            loaderScope.future {
                val comments = postService.getCommentsByPostIdIn(keys).toList()
                val mappedComments: MutableMap<UUID, List<Comment>> = mutableMapOf()
                keys.forEach { mappedComments[it] = comments.filter { c -> c.postId == it } }
                log.info("mapped comments: {}", mappedComments)
                mappedComments
            }
        }
    }
}